plugins/Synchro/SynchroSynth.cpp

```diff
@@ -0,0 +1,553 @@
+	{
+		STRINGIFY(PLUGIN_NAME),
+		"Synchro",
+		QT_TRANSLATE_NOOP("PluginBrowser", "2-oscillator PM synth"),
+		"Ian Sannar <ian/dot/sannar/at/gmail/dot/com>",
+		0x0100, //What does this do?
+		Plugin::Instrument,
+		new PluginPixmapLoader("logo"),
+		NULL,
+		NULL
+	};
+
+	PLUGIN_EXPORT Plugin * lmms_plugin_main(Model * m, void * data)
+	{
+		return new SynchroSynth(static_cast<InstrumentTrack *>(m));
+	}
+}
+
+//Keeps a phase value between 0 and 360
+static inline float wrapPhase(float phase) {
+	//Make sure phase is always between 0 and 2PI
+	while (phase >= F_2PI) { phase -= F_2PI; }
+	//Extra check since phase modulation can push it below 0
+	while (phase < 0) { phase += F_2PI; }
+	return phase;
+}
+
+//Triangle waveform generator
+//x: the phase of the waveform
+static inline float tri(float x)
+{
+	return 2.0 * fabs(2.0 * ((x / F_2PI) - floorf((x / F_2PI) + 0.5))) - 1.0;
+}
+
+//Triangle waveform generator with magic harmonics that make things dubsteppy
+//x: the phase of the waveform
+static inline float trih(float x, float harmonic)
+{
+	return tri(x) + ((tri(MAGIC_HARMONICS[0][0] * x) * MAGIC_HARMONICS[0][1]
+		+ tri(MAGIC_HARMONICS[1][0] * x) * MAGIC_HARMONICS[1][1])) * harmonic;
+}
+
+//Linear interpolation for envelopes
+//from: the initial value at x = 0
+//to: the final value at x = 1
+//x: the progress between the two values between 0 and 1
+static inline float lerp(float from, float to, float x)
+{
+	return from + x * (to - from);
+}
+
+//Exponential interpolation for envelopes
+//from: the initial value at x = 0
+//to: the final value at x = 1
+//x: the progress between the two values between 0 and 1
+static inline float expInterpol(float from, float to, float x)
+{
+	return from + fastSqrt(x) * (to - from);
+}
+
+//Detunes by octaves
+//pitch: the input pitch
+//detune: the amount of octaves to detune by
+static inline float DetuneOctaves(float pitch, float detune)
+{
+	return pitch * exp2(detune);
+}
+
+//Fast hyperbolic tangent approximation function
+//
+static inline float fastTanh(const float x)
+{
+	const double absX = fabs(x);
+	const double sqrX = x * x;
+	const double z = x * (0.773062670268356 + absX + (0.757118539838817 + 0.0139332362248817 * sqrX * sqrX) * sqrX * absX);
+	return z / (0.795956503022967 + fabs(z));
+}
+
+//Waveform function for the Synchro synthesizer.
+//x: input phase in radians (please keep it between 0 and 2PI)
+//drive: how much atan distortion is applied to the waveform
+//sync: hard sync with harmonic multiple
+//chop: how strong the amplitude falloff is per waveform period
+//harmonic: how strong the magic harmonics are
+static inline float SynchroWaveform(float x, float drive, float sync, float chop, float harmonic)
+{
+	return fastTanh(trih(F_PI_2 + x * sync, harmonic) * drive / powf(F_2PI / (F_2PI - x), chop));
+}
+
+//Envelope function
+static inline float ADSR(float attack, float decay, float sustain, float release, float time, bool isReleased, float releaseFramesDone) {
```

Comment:
`time` doesn't have a time dimension, so something like `framesPlayed` is more appropriate.
Also, could you add a comment that mentions ADR variables are in ms? You can rename the variables instead if you want.

Reply:
@PhysSong "@curlymorphic Any ideas how we can reduce aliasing while not losing f < Fs/2 bands significantly?" 

Oversampling could be the best solution and is commonly used, however, I would advise using a third-party library, as the filters will be optimised for this task as it is computationally expensive. The project I am currently on uses   https://github.com/xiph/speexdsp

Reply:
I'm not sure why I can't add comments to my original review. :thinking:
@curlymorphic Do you mean oversampling and then downsample it using 3rd-party library?

Reply:
The signal is being generated at the oversampled rate as expected. It is the downsampling that is the issue, just taking the mean, as the code currently does here is very simplistic.  Schematically the signal is generated at the oversampled rate,  then passed through a filter to remove content above the desired fs/2 then downsampled. This can be very computationally expensive.  A good overview can be found in "Designing audio effects plugin, 2nd, pirkle p624-640."

The filtering and downsampling code, could be written by LMMS. but a lot of development would be needed to make it both efficient and nice sounding, and I feel it would be a lot of effort just to reinvent the wheel.

Reply:
The [HIIR library by Laurent De Soras](http://ldesoras.free.fr/prod.html#src_hiir) was just added to LMMS as a 3rd party library. Maybe it could help here.